### PR TITLE
Verify TAS deployed properly with default values

### PR DESCRIPTION
**Pull request by jlojosnegros**
Improve verification of deployed resources when NRO use default values.
We need to check if all the resources has been properly deployed. 
There are multiple resources to check but we are going to start checking Deployments, DemonSets And CRDs

### CODE REVIEW

File: test/e2e/install/install_test.go [hunk 1]
```diff
@@ -58,8 +60,21 @@ var _ = Describe("[Install] continuousIntegration", func() {
 	})
 
 	Context("with a running cluster with all the components", func() {
-		It("should perform overall deployment and verify the condition is reported as available", func() {
-			deployedObj := overallDeployment()
+		var deployedObj []client.Object
+		BeforeEach(func() {
```
Comment (fromanirh) [inline test/e2e/install/install_test.go:64]: why was this moved to `beforeeach`?
I realize it could be confusing (and actually wrong) to leverage the side effect the suite is leaving TAS installed on the system; this should be at least documented.

File: test/e2e/install/install_test.go [hunk 2]
```diff
@@ -300,3 +329,30 @@ func getDaemonSetByOwnerReference(uid types.UID) (*v1.DaemonSet, error) {
...
+func getWorkerNodes() ([]*corev1.Node, error) {
+	nodeList := &corev1.NodeList{}
+	var ret []*corev1.Node
+
+	if err := e2eclient.Client.List(context.TODO(), nodeList); err != nil {
+		return nil, fmt.Errorf("failed to get nodes: %w", err)
+	}
+
+	for _, node := range nodeList.Items {
+		_, ok := node.Labels["node-role.kubernetes.io/worker"]
```
Comment (fromanirh) [inline test/e2e/install/install_test.go:351]: you can filter by labels using `ListOptions`.
This code is wrong because is too generic: we need to fetch the nodes referred by the `NodeGroups` we configured.
There is a fair bit of refactoring/code movement that can be useful for tests like this

File: test/e2e/sched/install/install_test.go [hunk 3]
```diff
@@ -71,9 +74,18 @@ var _ = Describe("[Scheduler] install", func() {
...
 			Expect(err).NotTo(HaveOccurred())
 			Expect(deploy.Status.ReadyReplicas).To(BeIdenticalTo(int32(1)))
+
+			By("checking the NumaResourcesScheduler CRD is deployed")
+			crd := &apiextensionv1.CustomResourceDefinition{}
+			key := client.ObjectKey{
+				Name: crdName,
+			}
+			err = e2eclient.Client.Get(context.TODO(), key, crd)
+			Expect(err).NotTo(HaveOccurred())
```
Comment (fromanirh) [inline test/e2e/sched/install/install_test.go:88]: I'm inclined to think we may need a new utility function. My own threshold is "used in two sites", and we already hit the threshold. More will likely come in the next batch of e2e tests.

File: test/e2e/sched/install/install_test.go [hunk 4]
```diff
@@ -71,9 +74,18 @@ var _ = Describe("[Scheduler] install", func() {
 			err = e2eclient.Client.Get(context.TODO(), client.ObjectKeyFromObject(nroSchedObj), nroSchedObj)
 			Expect(err).NotTo(HaveOccurred())
 
+			By("checking the NumaResourcesScheduler Deployment is correctly deployed")
 			deploy, err := schedutils.GetDeploymentByOwnerReference(nroSchedObj.UID)
 			Expect(err).NotTo(HaveOccurred())
 			Expect(deploy.Status.ReadyReplicas).To(BeIdenticalTo(int32(1)))
+
+			By("checking the NumaResourcesScheduler CRD is deployed")
```
Comment (fromanirh) [inline test/e2e/sched/install/install_test.go:82]: we also need to check the scheduler deployment

**Review by fromanirh — COMMENTED** [review]
we need more context both in the commit message and in the PR cover letter (the first message)

File: test/e2e/install/install_test.go (hunk 1 above)
Comment (fromanirh) [reply test/e2e/install/install_test.go:64]: documented -> not in this PR (separate one)

Comment (jlojosnegros) [reply test/e2e/install/install_test.go:64]: Well I just saw newly tests introduced in the package were following this pattern and I though it would be a good idea to make them all work in the same way and also clean-up the system after the test. 
I did not add any comment because I though it would not be needed as I was just using an already used pattern.
I will add some comments.

File: test/e2e/sched/install/install_test.go (hunk 4 above)
Comment (jlojosnegros) [reply test/e2e/sched/install/install_test.go:82]: Sorry but scheduler deployment is not the one checked in L78?

File: test/e2e/sched/install/install_test.go (hunk 3 above)
Comment (jlojosnegros) [reply test/e2e/sched/install/install_test.go:88]: ok

File: test/e2e/sched/install/install_test.go (hunk 4 above)
Comment (fromanirh) [reply test/e2e/sched/install/install_test.go:82]: right, overlooked this part

File: test/e2e/install/install_test.go (hunk 1 above)
Comment (fromanirh) [reply test/e2e/install/install_test.go:64]: ok, no problem. In this case we need a separate PR to document we intentionally leave the cluster state mutated after the install suite runs, and to explain why we do like we do.

Comment (jlojosnegros) [reply test/e2e/install/install_test.go:64]: Ok I though you want to comment why I was changing this to BeforeEach ... Now I see the point.

File: test/e2e/install/install_test.go [hunk 5]
```diff
@@ -38,12 +38,14 @@ import (
...
 	"github.com/openshift-kni/numaresources-operator/test/utils/configuration"
+	"github.com/openshift-kni/numaresources-operator/test/utils/crds"
 	"github.com/openshift-kni/numaresources-operator/test/utils/machineconfigpools"
 	"github.com/openshift-kni/numaresources-operator/test/utils/objects"
 )
 
 const (
-	crdName      = "noderesourcetopologies.topology.node.k8s.io"
+	crdNRTName   = "noderesourcetopologies.topology.node.k8s.io"
+	crdNROName   = "numaresourcesoperators.nodetopology.openshift.io"
```
Comment (fromanirh) [inline test/e2e/install/install_test.go:48]: not sure these belongs here or in the new `crds` package. But it's minor.

File: test/e2e/install/install_test.go [hunk 6]
```diff
@@ -90,12 +92,25 @@ var _ = Describe("[Install] continuousIntegration", func() {
...
 			Expect(err).NotTo(HaveOccurred())
+
+			By("checking Daemonset is up&running")
+			workerNodes, err := getWorkerNodes()
+			Expect(err).NotTo(HaveOccurred())
+
+			ds, err := getDaemonSetByOwnerReference(updatedNROObj.UID)
+			Expect(err).NotTo(HaveOccurred())
+
+			workerNodesNumber := int32(len(workerNodes))
```
Comment (fromanirh) [inline test/e2e/install/install_test.go:109]: rather than using this value, we should use the `desiredNumberScheduled` as target. We need to assume the value is correctly computed (and maybe test it on a separate test). IOW, we don't need to count the worker nodes (which we are doing incorrectly here).

File: test/e2e/install/install_test.go [hunk 7]
```diff
@@ -90,12 +92,25 @@ var _ = Describe("[Install] continuousIntegration", func() {
...
+			Expect(ds.Status.DesiredNumberScheduled).To(Equal(workerNodesNumber))
+			Expect(ds.Status.CurrentNumberScheduled).To(Equal(workerNodesNumber))
+			Expect(ds.Status.NumberReady).To(Equal(workerNodesNumber))
```
Comment (fromanirh) [inline test/e2e/install/install_test.go:112]: this works in the happy path: lines 106-112 should be wrapped in a `Eventually` block.

File: test/e2e/sched/install/install_test.go [hunk 8]
```diff
@@ -71,9 +74,14 @@ var _ = Describe("[Scheduler] install", func() {
 			err = e2eclient.Client.Get(context.TODO(), client.ObjectKeyFromObject(nroSchedObj), nroSchedObj)
 			Expect(err).NotTo(HaveOccurred())
 
+			By("checking the NumaResourcesScheduler Deployment is correctly deployed")
 			deploy, err := schedutils.GetDeploymentByOwnerReference(nroSchedObj.UID)
 			Expect(err).NotTo(HaveOccurred())
 			Expect(deploy.Status.ReadyReplicas).To(BeIdenticalTo(int32(1)))
```
Comment (fromanirh) [inline test/e2e/sched/install/install_test.go:80]: just noticed that this check has the same issue of the daemonset check. Can be fixed in a later PR.

File: test/e2e/install/install_test.go (hunk 5 above)
Comment (jlojosnegros) [reply test/e2e/install/install_test.go:48]: yeah you are totally right ... will change it

File: test/e2e/install/install_test.go (hunk 6 above)
Comment (jlojosnegros) [reply test/e2e/install/install_test.go:109]: Ok, understood. I will change it.

For the sake of curiosity... will this be a proper way to found the nodes where daemonset should be running ?

```go
func getWorkerNodes(labelSelector map[string]string) ([]corev1.Node, error) {
	nodeList := &corev1.NodeList{}

	if err := e2eclient.Client.List(context.TODO(), nodeList, &client.ListOptions{LabelSelector: labels.SelectorFromSet(labelSelector)}); err != nil {
		return nil, fmt.Errorf("failed to get nodes: %w", err)
	}
	return nodeList.Items, nil
}
```
using  the MachineConfigPool.Spec.NodeSelector.MatchLabels as `labelSelector`

File: test/e2e/install/install_test.go (hunk 7 above)
Comment (jlojosnegros) [reply test/e2e/install/install_test.go]: done

File: test/e2e/install/install_test.go [hunk 9]
```diff
@@ -90,12 +89,45 @@ var _ = Describe("[Install] continuousIntegration", func() {
...
+					klog.Infof(" CurrentNumberScheduled %d (should be %d)", ds.Status.CurrentNumberScheduled, ds.Status.DesiredNumberScheduled)
+					return false
+				}
+
+				if ds.Status.NumberReady != ds.Status.DesiredNumberScheduled {
+					klog.Infof(" NumberReady %d (should be %d)", ds.Status.CurrentNumberScheduled, ds.Status.DesiredNumberScheduled)
+					return false
+				}
+				return true
+			}, DSCheckTimeout, DSCheckPollingPeriod).Should(BeTrue(), "DaemonSet Status was not correct")
```
Comment (fromanirh) [inline test/e2e/install/install_test.go:130]: for the next time: "DaemonSet status is not correct" is too generic. The test should clearly signal at glance what the expectation is and what the actual state is. Yes, we (correctly) do logging, but in some test runners we may want not to enable the logs by default, so better to reiterate the failed condition as precisely as possible.

**Review by fromanirh — COMMENTED** [review]
/lgtm
/approve
we will need a followup to move the deployment verification in Eventually

**Review by fromanirh — COMMENTED** [review]
/lgtm

File: test/e2e/install/install_test.go (hunk 6 above)
Comment (fromanirh) [reply test/e2e/install/install_test.go]: there is code in the `controllers` package (soon to be moved on its own package) to resolve nodegroups which we can use here
